[SPARK-26755][SCHEDULER] : Optimize Spark Scheduler to dequeue speculative tasks…

pgandhi999 opened the pull request:
… more efficiently

This PR improves the performance of scheduling speculative tasks to be O(1) instead of O(numSpeculativeTasks), using the same approach used for scheduling regular tasks. The performance of this method is particularly important because a lock is held on the TaskSchedulerImpl which is a bottleneck for all scheduling operations. We ran a Join query on a large dataset with speculation enabled and out of 100000 tasks for the ShuffleMapStage, the maximum number of speculatable tasks that was noted was close to 7900-8000 at a point. That is when we start seeing the bottleneck on the scheduler lock.

In particular, this works by storing a separate stack of tasks by executor, node, and rack locality preferences. Then when trying to schedule a speculative task, rather than scanning all speculative tasks to find ones which match the given executor (or node, or rack) preference, we can jump to a quick check of tasks matching the resource offer. This technique was already used for regular tasks -- this change refactors the code to allow sharing with regular and speculative task execution.

## What changes were proposed in this pull request?

Have split the main queue "speculatableTasks" into 5 separate queues based on locality preference similar to how normal tasks are enqueued. Thus, the "dequeueSpeculativeTask" method will avoid performing locality checks for each task at runtime and simply return the preferable task to be executed.

## How was this patch tested?
We ran a spark job that performed a join on a 10 TB dataset to test the code change.
Original Code:
<img width="1433" alt="screen shot 2019-01-28 at 5 07 22 pm" src="https://user-images.githubusercontent.com/22228190/51873321-572df280-2322-11e9-9149-0aae08d5edc6.png">


Optimized Code:
<img width="1435" alt="screen shot 2019-01-28 at 5 08 19 pm" src="https://user-images.githubusercontent.com/22228190/51873343-6745d200-2322-11e9-947b-2cfd0f06bcab.png">


As you can see, the run time of the ShuffleMapStage came down from 40 min to 6 min approximately, thus, reducing the overall running time of the spark job by a significant amount.

Another example for the same job:

Original Code:
<img width="1440" alt="screen shot 2019-01-28 at 5 11 30 pm" src="https://user-images.githubusercontent.com/22228190/51873355-70cf3a00-2322-11e9-9c3a-af035449a306.png">


Optimized Code:
<img width="1440" alt="screen shot 2019-01-28 at 5 12 16 pm" src="https://user-images.githubusercontent.com/22228190/51873367-7dec2900-2322-11e9-8d07-1b1b49285f71.png">

Commit:
[SPARK-26755] : Optimize Spark Scheduler to dequeue speculative tasks more efficiently

Have split the main queue "speculatableTasks" into 5 separate queues based on locality preference similar to how normal tasks are enqueued.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

pgandhi999 commented:
ok to test

SparkQA commented:
**[Test build #101775 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101775/testReport)** for PR 23677 at commit [`2d87a62`](https://github.com/apache/spark/commit/2d87a62438e8f3a3267bcd7ab15a213a98ca61eb).

pgandhi999 commented:
cc @tgravescs @vanzin @abellina @redsanket @dhruve

SparkQA commented:
**[Test build #101775 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101775/testReport)** for PR 23677 at commit [`2d87a62`](https://github.com/apache/spark/commit/2d87a62438e8f3a3267bcd7ab15a213a98ca61eb).
 * This patch **fails from timeout after a configured wait of `400m`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101775/
Test FAILed.

abellina commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -1029,12 +1062,11 @@ private[spark] class TaskSetManager(
-        if (!successful(index) && copiesRunning(index) == 1 && info.timeRunning(time) > threshold &&
-          !speculatableTasks.contains(index)) {
+        if (!successful(index) && copiesRunning(index) == 1 && info.timeRunning(time) > threshold) {
+          addPendingSpeculativeTask(index)
```
In this check we ensure 1 task is running (so it's been dequeued). While the 1 task is running, and we mark it for potential speculation, we are not protecting anymore against duplicates. So each of the queues could have duplicate tasks to speculate.

Could this lead to multiple speculative tasks for a single real task? And also, could it lead to a snowball effect where speculative tasks are created off of speculative tasks that are running? Seems like the previous code was protecting against that in `!speculatableTasks.contains(index)`.

pgandhi999 replied:
Yes, you were right, have updated the code to use HashSet instead of ArrayBuffer. Thank you.

squito replied:
can you also add a test case that if you have multiple calls to checkSpeculatbleTasks followed by multiple resources offers, you still only submit one speculative task?

pgandhi999 replied:
Have added the test case.

redsanket commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -245,6 +257,28 @@ private[spark] class TaskSetManager(
+    }
+
+    // No point scanning this whole list to find the old task there
+    allPendingSpeculatableTasks += index
```
qq does this need to hold all tasks or only with TaskLocality.ANY pref... seems like a waste to scan all the tasks within it for getting non local tasks, could be wrong in my understanding however thanks

redsanket replied:
Also in your tests would be interesting to note what is locality affinity... do you see more process local tasks vs node local etc., that might be related to scheduling logic, just curious to know about it.

pgandhi999 replied:
So afaik, task locality here means a preference for the individual task to run at a desired locality level(kind of like for flight scheduling, my preference would be to fly from airport A as it is the nearest to my house, but if the flight i want to book does not take off from location A, location B which is farther than A, will do). If the desired locality level cannot be assigned to the task by the scheduler due to locality constraints and other checks, it will be assigned the next best locality. That is why we need all tasks in the ANY locality preference queue. The method `hasAttemptOnHost` will take care of duplicate scheduling of tasks anyways.

pgandhi999 replied:
Good point. Will definitely test for locality affinity, but ideally, that logic should not have changed with this PR.

Commit:
[SPARK-26755] : Using HashSet instead of ArrayBuffer to reject duplicate task index

SparkQA commented:
**[Test build #101833 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101833/testReport)** for PR 23677 at commit [`65d0926`](https://github.com/apache/spark/commit/65d09265bc6ae16cf595798e2a24e3d61c0ff0e9).

squito commented:
I haven't looked very closely yet, but general idea makes sense, and results are very compelling.

High level question though -- any reason `TaskSetManager.canFetchMoreResults` needs to lock the `TaskSchedulerImpl`?  Couldn't we protect `tsm.totalResultSize` and `tsm.calculatedTasks` with something local to the `tsm`, which would be entirely separate from the lock needed for speculation?

I'm really just brainstorming at this point -- that change might actually be a lot harder to reason about.

SparkQA commented:
**[Test build #101833 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101833/testReport)** for PR 23677 at commit [`65d0926`](https://github.com/apache/spark/commit/65d09265bc6ae16cf595798e2a24e3d61c0ff0e9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101833/
Test PASSed.

pgandhi999 commented:
@squito That is a good point to look into. The lock at the task scheduler level was introduced to  `TaskSetManager.canFetchMoreResults` in https://github.com/apache/spark/commit/c306555f491e45ef870f58938af397f9ec5f166a with the corresponding JIRA https://issues.apache.org/jira/browse/SPARK-5219. The JIRA is not descriptive enough so I do not know the exact reason as to why was the lock added.

SparkQA commented:
**[Test build #102999 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102999/testReport)** for PR 23677 at commit [`65d0926`](https://github.com/apache/spark/commit/65d09265bc6ae16cf595798e2a24e3d61c0ff0e9).

SparkQA commented:
**[Test build #102999 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102999/testReport)** for PR 23677 at commit [`65d0926`](https://github.com/apache/spark/commit/65d09265bc6ae16cf595798e2a24e3d61c0ff0e9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102999/
Test PASSed.

pgandhi999 commented:
Hi @squito , any chance we could take a look at this one. We have experimented quite a lot with this code change and we have seen quite positive results especially in terms of running time improvements. Thank you.

Commit:
Merge branch 'master' of https://github.com/pgandhi999/spark into SPARK-26755

pgandhi999 commented:
@tgravescs @vanzin @squito We have been running Spark on our clusters with this code patch and have observed runtime improvements for large jobs by a factor of 20-25 mins. Would appreciate your reviews. Thank you.

SparkQA commented:
**[Test build #106899 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106899/testReport)** for PR 23677 at commit [`e525b1c`](https://github.com/apache/spark/commit/e525b1c21940f2b24c062b1cf5ddee4a5b7e9212).

SparkQA commented:
**[Test build #106899 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106899/testReport)** for PR 23677 at commit [`e525b1c`](https://github.com/apache/spark/commit/e525b1c21940f2b24c062b1cf5ddee4a5b7e9212).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106899/
Test FAILed.

pgandhi999 commented:
test this please

SparkQA commented:
**[Test build #106902 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106902/testReport)** for PR 23677 at commit [`e525b1c`](https://github.com/apache/spark/commit/e525b1c21940f2b24c062b1cf5ddee4a5b7e9212).

SparkQA commented:
**[Test build #106902 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106902/testReport)** for PR 23677 at commit [`e525b1c`](https://github.com/apache/spark/commit/e525b1c21940f2b24c062b1cf5ddee4a5b7e9212).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106902/
Test PASSed.

squito commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -158,9 +158,21 @@ private[spark] class TaskSetManager(
+  private[scheduler] var pendingSpeculatableTasksForRack = new HashMap[String, HashSet[Int]]
+
+  // Set of all pending tasks that can be speculated.
+  private[scheduler] val allPendingSpeculatableTasks = new HashSet[Int]
```
overall, this makes a lot of sense.  I see how you're replacing a bunch of `O(numSpeculatableTasks)` operations with things specific to tasks targeted at each executor, etc., just like the other scheduler methods.

I don't really like how there is so much code duplication between the speculative and non-speculative versions, for what seem like minor differences.  I was thinking you could create a private case class like

```scala
case class PendingsTasksByLocality(forExecutor: HashMap[String, ArrayBuffer[Int]], forHost: HashMap[String, ArrayBuffer[Int]], ...)
```

with one instance for all pending tasks, one for pending speculative tasks.  Then a lot of the functions could be combined (eg. dequeueSpeculativeTaskFromList & dequeueTaskFromList) just with a `speculative: Boolean` parameter added.  Your new datastructures are `HashMap[String, HashSet]` instead of `HashMap[String, ArrayBuffer]`, but I think you can go back to `ArrayBuffer` if you *also* keep `val speculatableTasks = new HashSet[Int]` for avoiding adding duplicate entries.

pgandhi999 replied:
makes sense @squito, I have refactored the code. Please let me know if it looks different from what you had thought earlier. Thank you.

squito reviewed (commented):
sorry its taken me so long to get back to this @pgandhi999 .  Makes a lot of sense to me, just wondering if we can clean things up some, improve tests etc.

Commit:
[SPARK-26755] : Addressing Reviews July 1, 2019

Adding Unit test, refactoring code to remove duplicate methods, refactoring comments etc.

SparkQA commented:
**[Test build #107083 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107083/testReport)** for PR 23677 at commit [`41ddf23`](https://github.com/apache/spark/commit/41ddf23f96b11afd3468fe358b4c16b3d4622d70).

pgandhi999 commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -269,6 +281,28 @@ private[spark] class TaskSetManager(
+        pendingSpeculatableTasksWithNoPrefs += index
+    }
+
+    // No point scanning this whole list to find the old task there
```
Done

pgandhi999 commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -318,6 +352,29 @@ private[spark] class TaskSetManager(
+  private def dequeueSpeculativeTaskFromList(
+    execId: String,
+    host: String,
+    list: HashSet[Int]): Option[Int] = {
```
Done

SparkQA commented:
**[Test build #107083 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107083/testReport)** for PR 23677 at commit [`41ddf23`](https://github.com/apache/spark/commit/41ddf23f96b11afd3468fe358b4c16b3d4622d70).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class PendingTasksByLocality(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107083/
Test PASSed.

squito commented on core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:
```diff
@@ -1655,4 +1657,81 @@ class TaskSetManagerSuite extends SparkFunSuite with LocalSparkContext with Logg
+    // allPendingSpeculativeTasks will still have two pending tasks but
+    // pendingSpeculatableTasksWithNoPrefs should have none
+    assert(manager.pendingSpeculatableTasks.anyPrefs.size === 2)
+    assert(manager.pendingSpeculatableTasks.noPrefs.size === 0)
```
these structures are actually an implementation detail, in particular since they leave some "stale" state and are meant to lazily correct themselves.  So I wouldn't test their size at all.

However, here you *should* test what happens if you have another resource offer, with `ANY` locality, and make sure you don't schedule two tasks on them.  Furthermore, you should probably originally give your tasks some locality preferences, and then make offers with some meeting the locality preferences, and some not meeting the locality constraints.  Since you'll have the speculatable task listed in multiple places inside `pendingSpeculatableTasks`, we want to make sure we don't run multiple copies.

pgandhi999 replied:
@squito Your comment does make sense and I can implement it in the test. The only part I am stuck at is trying to figure out a way to test that we are not resubmitting speculative tasks without checking in the HashMap or the size. Would appreciate your guidance in this matter. Thank you.

squito replied:
Just make more resource offers, and see if they result in anything getting scheduled.  Be sure it includes an offer on a different host, to avoid the `canRunOnHost` filter.  Eg. like this (lots of extra logs in there):

https://github.com/squito/spark/commit/6cdd8ee82337d223f49d9a43bf7295b963720d4d

my added test *fails* because that last offer, on host3, results in another speculative task.  I think we need a filter w/ speculative execution that there is <= 2 tasks running.

pgandhi999 replied:
@squito I have modified the test. Thank you for pointing me out in the right direction. It should more or less look good now, please let me know if it still needs modification. Thank you.

Commit:
Merge branch 'master' of https://github.com/pgandhi999/spark into SPARK-26755

[SPARK-26755] : Upmerging with master branch

Commit:
[SPARK-26755] : Addressing Reviews July 8, 2019

Restructuring code and eliminating duplicate code

pgandhi999 commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -143,36 +144,25 @@ private[spark] class TaskSetManager(
+    forHost = new HashMap[String, ArrayBuffer[Int]],
+    noPrefs = new ArrayBuffer[Int],
+    forRack = new HashMap[String, ArrayBuffer[Int]],
+    anyPrefs = new ArrayBuffer[Int])
```
Done

pgandhi999 commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -246,85 +236,86 @@ private[spark] class TaskSetManager(
-        case _ =>
+      resolveRacks: Boolean = true,
+      speculative: Boolean = false): Unit = {
+    if (speculative) {
```
Done

pgandhi999 commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -336,37 +327,30 @@ private[spark] class TaskSetManager(
-          list -= index
-          if (!successful(index)) {
+          list.remove(indexOffset)
+          if (copiesRunning(index) == 0 && !successful(index)) {
```
Done

pgandhi999 commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -1095,11 +1083,13 @@ private[spark] class TaskSetManager(
-        if (!successful(index) && copiesRunning(index) == 1 && info.timeRunning(time) > threshold) {
-          addPendingSpeculativeTask(index)
+        if (!successful(index) && copiesRunning(index) == 1 && info.timeRunning(time) > threshold &&
+          !speculatableTasks.contains(index)) {
```
Done

pgandhi999 commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -246,85 +236,86 @@ private[spark] class TaskSetManager(
    */
   private def getPendingTasksForExecutor(executorId: String): ArrayBuffer[Int] = {
-    pendingTasksForExecutor.getOrElse(executorId, ArrayBuffer())
+    pendingTasks.forExecutor.getOrElse(executorId, ArrayBuffer())
```
Done. Removed the methods.

pgandhi999 commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -398,23 +382,25 @@ private[spark] class TaskSetManager(
-      execId, host, pendingSpeculatableTasksForExecutor.getOrElse(execId, HashSet()))) {
+    for (index <- dequeueTaskFromList(
+      execId, host, pendingSpeculatableTasks.forExecutor.getOrElse(execId, ArrayBuffer()),
+      speculative = true)) {
```
Done. Thank you for your valuable help in coming up with the code.

SparkQA commented:
**[Test build #107361 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107361/testReport)** for PR 23677 at commit [`1fa17ec`](https://github.com/apache/spark/commit/1fa17ec35b8320c0c627b27abd650d75946d5c72).

SparkQA commented:
**[Test build #107361 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107361/testReport)** for PR 23677 at commit [`1fa17ec`](https://github.com/apache/spark/commit/1fa17ec35b8320c0c627b27abd650d75946d5c72).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107361/
Test FAILed.

Commit:
[SPARK-26755] : Fixing Scalastyle test failures

SparkQA commented:
**[Test build #107362 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107362/testReport)** for PR 23677 at commit [`025e548`](https://github.com/apache/spark/commit/025e54840e69c030cb153d42aab316a942360cff).

SparkQA commented:
**[Test build #107362 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107362/testReport)** for PR 23677 at commit [`025e548`](https://github.com/apache/spark/commit/025e54840e69c030cb153d42aab316a942360cff).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107362/
Test PASSed.

Commit:
[SPARK-26755] : Addressing Reviews July 9, 2019

Modifying test, adding extra condition to not let more than two speculative tasks being spawned

pgandhi999 commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -371,107 +304,64 @@ private[spark] class TaskSetManager(
+    execId: String,
+    host: String,
+    maxLocality: TaskLocality.Value,
+    speculative: Boolean): Option[(Int, TaskLocality.Value, Boolean)] = {
```
Done

pgandhi999 commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -238,84 +228,41 @@ private[spark] class TaskSetManager(
-            pendingTasks.forRack.getOrElseUpdate(rack, new ArrayBuffer) += index
+    val pendingTaskSetToAddTo = if (speculative) pendingSpeculatableTasks else pendingTasks
+    // ... mostly the original code from `addPendingTask` here, just adding
+    // into pendingTaskSetToAddTo
```
Done, removed the comment

pgandhi999 commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -371,107 +304,64 @@ private[spark] class TaskSetManager(
-    for (index <- dequeueTaskFromList(execId, host, getPendingTasksForExecutor(execId))) {
-      return Some((index, TaskLocality.PROCESS_LOCAL, false))
+  : Option[(Int, TaskLocality.Value, Boolean)] = {
+    // if we didn't schedule a regular task, try to schedule a speculative one
```
makes sense, have modified the comment.

SparkQA commented:
**[Test build #107414 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107414/testReport)** for PR 23677 at commit [`b58c34f`](https://github.com/apache/spark/commit/b58c34f7ed7c0feb612818560a156a41823ab566).

SparkQA commented:
**[Test build #107414 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107414/testReport)** for PR 23677 at commit [`b58c34f`](https://github.com/apache/spark/commit/b58c34f7ed7c0feb612818560a156a41823ab566).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107414/
Test PASSed.

squito commented:
I need to do another careful pass, which will probably not happen till next week, but I think this probably ready.  Would like to ping a couple more reviewers if they'd like to take a look in the meantime @tgravescs @Ngone51 @jiangxb1987

Ngone51 reviewed (commented):
The general idea lgtm and the performance improvement is really convincing. Just leaves some comments.

squito commented on core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:
```diff
@@ -1655,4 +1657,70 @@ class TaskSetManagerSuite extends SparkFunSuite with LocalSparkContext with Logg
     assert(availableResources(GPU) sameElements Array("0", "1", "2", "3"))
   }
+
+  test("SPARK-26755 Ensure that a speculative task is submitted only once for execution") {
```
no, it was correct before, but the change in implementation is susceptible to this bug (early versions of this PR had this bug)

Ngone51 replied:
I see, thanks for reminding that.

squito commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -330,128 +303,65 @@ private[spark] class TaskSetManager(
-      for (index <- dequeueTaskFromList(execId, host, pendingTasksWithNoPrefs)) {
-        return Some((index, TaskLocality.PROCESS_LOCAL, false))
+      dequeue(pendingTaskSetToUse.noPrefs).foreach { index =>
+        return Some((index, TaskLocality.NO_PREF, speculative))
```
hmm, this is a very good point.  I traced this back to to https://issues.apache.org/jira/browse/SPARK-2294 / https://github.com/apache/spark/commit/63bdb1f41b4895e3a9444f7938094438a94d3007 /  https://github.com/apache/spark/pull/1313 (loooooooots of comments).

Changing this would affect the `currentLocalityIndex` here: https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L505-L507

and also will make its way into the UI related changes here:
https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala#L69

(and here, but this will probably be fine: https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala#L608)

I am not sure whether the old behavior is right or not, but in any case I think we should not change that as part of this change, and should consider updating it separately.

pgandhi999 replied:
Hmm I see, have updated the code patch, let me know if it still looks logically incorrect.

Ngone51 replied:
@squito Thanks for attaching those links. They're useful to figure out what's the right behavior. But before it, remaining the old behavior is fine.

Commit:
[SPARK-26755]: Merge branch 'master' of https://github.com/pgandhi999/spark into SPARK-26755

Upmerging with master branch

Commit:
[SPARK-26755] : Addressing Reviews July 12, 2019

Refactoring comments, changing NO_PREF to PROCESS_LOCAL in code etc.

pgandhi999 commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -302,15 +272,18 @@ private[spark] class TaskSetManager(
         list.remove(indexOffset)
-        if (copiesRunning(index) == 0 && !successful(index)) {
+        if (((copiesRunning(index) < 2 && speculative) ||
+          (!speculative && copiesRunning(index) == 0)) && !successful(index)) {
```
Makes sense, updated code.

pgandhi999 commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -1137,3 +1048,12 @@ private[spark] object TaskSetManager {
+  val forHost: HashMap[String, ArrayBuffer[Int]] = new HashMap[String, ArrayBuffer[Int]]
+  val noPrefs: ArrayBuffer[Int] = new ArrayBuffer[Int]
+  val forRack: HashMap[String, ArrayBuffer[Int]] = new HashMap[String, ArrayBuffer[Int]]
+  val anyPrefs: ArrayBuffer[Int] = new ArrayBuffer[Int]
```
Done

pgandhi999 commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -143,25 +144,16 @@ private[spark] class TaskSetManager(
-  private val pendingTasksForExecutor = new HashMap[String, ArrayBuffer[Int]]
 
-  // Set of pending tasks for each host. Similar to pendingTasksForExecutor,
-  // but at host level.
```
Done

pgandhi999 commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -143,25 +144,16 @@ private[spark] class TaskSetManager(
-  private val allPendingTasks = new ArrayBuffer[Int]
-
-  // Tasks that can be speculated. Since these will be a small fraction of total
-  // tasks, we'll just hold them in a HashSet.
```
Done

SparkQA commented:
**[Test build #107616 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107616/testReport)** for PR 23677 at commit [`62a15d7`](https://github.com/apache/spark/commit/62a15d7eb17b21378448e6ab2277434285e75ba1).

SparkQA commented:
**[Test build #107616 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107616/testReport)** for PR 23677 at commit [`62a15d7`](https://github.com/apache/spark/commit/62a15d7eb17b21378448e6ab2277434285e75ba1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107616/
Test PASSed.

Ngone51 commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -143,25 +144,18 @@ private[spark] class TaskSetManager(
-  // tasks, we'll just hold them in a HashSet.
+  // tasks, we'll just hold them in a HashSet. The HashSet here ensures that we do not add
+  // duplicate speculative tasks.
   private[scheduler] val speculatableTasks = new HashSet[Int]
```
I see, now, this is only used to guard against submitting duplicate speculatable tasks, right ? But it used to dequeue tasks by traversing on it for several locality level. And this why it slows down the job when it is large. 

But from the comment above, I think we have a default agreement previously that speculatable tasks should always be less in most cases(even for large taskset). And this maybe the reason why we don't classify it by location pref as non-speculatable tasks do.

So, just out of curiosity, my question is, about how many speculatable tasks spawn out of the stage(e.g. 100000 partitions) ? And(assuming the cluster is healthy), why there's so many  speculatable tasks ? Do you have any specific configs ? I'm wondering that current speculation mechanism may have something wrong if nothing's abnormal in this case.

pgandhi999 replied:
@Ngone51 that is a valid question indeed. This change was reasoned about due to some jobs we saw on the cluster in which speculation was heavily kicking in due to the fact that the cluster during those points in time **was healthy but was really busy**. As I have documented the findings in https://issues.apache.org/jira/browse/SPARK-26755, you can see that the task-result-getter threads were more than often in BLOCKED state and job progress was slow.
 The HashSet `speculatableTasks` was still holding a small fraction of the tasks(at most 10 to 15%) but every call to `dequeueSpeculativeTask` would end up looping through each task 5 different times in the worst case thus, leading to unnecessary traversing and thread blocks.

squito replied:
I think you've given some good evidence, its easy to see how anything `O(numSpeculatableTasks)` on every resource offer would lead to really poor performance.  But it would be great if you could share the size of the taskset & the numSpeculatableTasks (probably easiest if you add a logline in checkSpeculatableTasks here, just for your internal repro: https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L1071)

I think this hasn't been handled before just because speculative execution & large tasks just hasn't been looked at that closely before.

pgandhi999 replied:
So I ran a Join query on a dataset of size 10TB without this change and out of 100000 tasks for the ShuffleMapStage, the maximum number of speculatable tasks that was noted was close to 7900-8000 at a point. That is when we start seeing the bottleneck on the scheduler lock.

Ngone51 replied:
@pgandhi999 Thanks for doing the experiment and posting the details. I think they're really helpful for further digs.

pgandhi999 replied:
Thank you @Ngone51 once again for your valuable reviews and suggestions. Really appreciate it!

squito commented on core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:
```diff
@@ -1655,4 +1657,70 @@ class TaskSetManagerSuite extends SparkFunSuite with LocalSparkContext with Logg
+    assert(manager.copiesRunning(2) === 1)
+    assert(manager.copiesRunning(3) === 1)
+
+    // Offer resource to start the speculative attempt for the running task
```
Lets expand this comment a bit to explain that we're about to start the main point of this test, something like.

"We offer more resources, and ensure that speculative tasks get scheduled appropriately -- one extra copy per speculatable task, and still obeying the original locality preferences"


and writing that extended description makes me realize we are actually *not* checking locality that preferences are obeyed for speculative tasks in this test.  To do that, I think you need to expand this a bit:

* the tasks should have locality preferences, say hosts 1 & 3 for task 1, and host 2 for task 2
* add a exec4 / host4 (with no task that prefers that host)
* after you've got speculatable tasks, first make an offer with max locality NODE_LOCAL but on exec4, check it does *not* get scheduled
* make an offer on exec2, still with NODE_LOCAL preferences -- task does *not* get scheduled because of "same host" exclusion for speculative tasks
* make an offer on exec3, still with NODE_LOCAL -- task *does* get scheduled, because its a different host than running task1 and falls within locality prefs
* make an offer on exec4 with ANY max locality, and task2 should get a speculative copy scheduled there.
* make more offers, nothing scheduled (because copiesRunning == 2 for all tasks already)
* another call to checkSpeculatable, and more offers, still nothing scheduled

pgandhi999 replied:
Have written a separate unit test to account for the locality preference feature. Thank you.

squito reviewed (commented):
Hi @pgandhi999 some style / comment changes and a request to expand the test.

Thanks again for the work on this, and sorry for the delayed review.  just fyi I'll try to be responsive this week, but I'll be out next week -- would love to wrap this up by then.

squito commented:
Can you also update the PR description, to something which just focuses more on the actual improvement and less on the discovery of the issue?   I think you should just mention something like

"This improves the performance of scheduling speculative tasks to be O(1) instead of O(numSpeculativeTasks), using the same approach used for scheduling regular tasks.  The performance of this method is particularly important because a lock is held on the TaskSchedulerImpl which is a bottleneck for all scheduling operations.

In particular, this works by storing a separate stack of tasks by executor, node, and rack locality preferences.  Then when trying to schedule a speculative task, rather than scanning *all* speculative tasks to find ones which match the given executor (or node, or rack) preference, we can jump to a quick check of tasks matching the resource offer.  This technique was already used for regular tasks -- this change refactors the code to allow sharing with regular and speculative task execution."

squito commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -1064,7 +979,8 @@ private[spark] class TaskSetManager(
+          addPendingTask(index, speculative = true)
           logInfo(
             "Marking task %d in stage %s (on %s) as speculatable because it ran more than %.0f ms"
               .format(index, taskSet.id, info.host, threshold))
```
thinking a bit more about my comment that you should change this logline ... if you have a ton of speculative tasks, isn't this horribly verbose?  I am wondering if it should be dropped to logDebug.  But then I guess we lose all info about speculatable assignments, which would be a bummer.  But at the very least, since this comment is here, its probably worth also throwing in the total number of speculatable tasks?  "... ran more than 100 ms (145 speculatable tasks in this taskset now)"

pgandhi999 replied:
Makes sense, have updated the comment.

Commit:
[SPARK-26755] : Addressing Reviews July 16, 2019

Adding a unit test to check for locality prefs for speculatable tasks and other changes

SparkQA commented:
**[Test build #107762 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107762/testReport)** for PR 23677 at commit [`466849e`](https://github.com/apache/spark/commit/466849e3ebf8e39f994423947f21bcf390c26894).

pgandhi999 commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -330,128 +308,65 @@ private[spark] class TaskSetManager(
-  {
-    for (index <- dequeueTaskFromList(execId, host, getPendingTasksForExecutor(execId))) {
-      return Some((index, TaskLocality.PROCESS_LOCAL, false))
+  : Option[(Int, TaskLocality.Value, Boolean)] = {
```
Done

pgandhi999 commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -143,25 +144,18 @@ private[spark] class TaskSetManager(
   // of failures.
   // Duplicates are handled in dequeueTaskFromList, which ensures that a
   // task hasn't already started running before launching it.
```
Done

pgandhi999 commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -143,25 +144,18 @@ private[spark] class TaskSetManager(
   // Tasks that can be speculated. Since these will be a small fraction of total
-  // tasks, we'll just hold them in a HashSet.
+  // tasks, we'll just hold them in a HashSet. The HashSet here ensures that we do not add
+  // duplicate speculative tasks.
```
Done

pgandhi999 commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -234,63 +228,41 @@ private[spark] class TaskSetManager(
       index: Int,
-      resolveRacks: Boolean = true): Unit = {
+      resolveRacks: Boolean = true,
+      speculative: Boolean = false): Unit = {
```
Done

pgandhi999 commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -302,16 +274,22 @@ private[spark] class TaskSetManager(
       val index = list(indexOffset)
-      if (!isTaskBlacklistedOnExecOrNode(index, execId, host)) {
+      if (!isTaskBlacklistedOnExecOrNode(index, execId, host) &&
+        !(speculative && hasAttemptOnHost(index, host))) {
```
Done

pgandhi999 commented on core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:
```diff
@@ -1655,4 +1657,70 @@ class TaskSetManagerSuite extends SparkFunSuite with LocalSparkContext with Logg
+      "exec1" -> "host1",
+      "exec1" -> "host1",
+      "exec2" -> "host2",
+      "exec2" -> "host2")) {
```
Done

pgandhi999 commented:
@squito @Ngone51 Have updated the PR with the changes. Thank you once again for your valuable reviews.

squito commented on core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:
```diff
@@ -1723,4 +1722,48 @@ class TaskSetManagerSuite extends SparkFunSuite with LocalSparkContext with Logg
+    assert(manager.resourceOffer("exec4", "host4", NODE_LOCAL).isEmpty)
+    assert(manager.resourceOffer("exec2", "host2", NODE_LOCAL).isEmpty)
+    assert(manager.resourceOffer("exec3", "host3", NODE_LOCAL).isDefined)
+    assert(manager.resourceOffer("exec4", "host4", ANY).isDefined)
```
any particular reason to pull this out into a separate test case?  Seems like it could be combined.  Its fine if there is a good reason, but I don't like a proliferation of test cases that are all doing more or less the same thing.  It seems the only thing which you aren't doing here, but you are doing above, is checking the taskId etc. of the speculative tasks.

also another thing missing from both tests -- there is no check that we do not schedule a speculative task on the same host as the original task, even despite locality preferences.

(I realize some of these tests were missing before, but this logic is getting a little trickier now, and maybe those tests always should have been there)

pgandhi999 replied:
@squito I have combined the two test cases into one. Also have added the missing test(checking whether speculatable task is launched on same host as original). Please let me know if anything else needs to be added/removed from the test. Thank you.

squito reviewed (commented):
Can you just remove the first part of the PR description, before the updated part I suggested?  That is all good info to have in the jira, but I don't think its useful in the commit msg.

Also can you update the testing section above to mention the number of tasks you mentioned above?  More important than the size of the dataset is that you ran with 100K tasks and ~8K speculatable tasks.

SparkQA commented:
**[Test build #107762 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107762/testReport)** for PR 23677 at commit [`466849e`](https://github.com/apache/spark/commit/466849e3ebf8e39f994423947f21bcf390c26894).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107762/
Test PASSed.

Commit:
[SPARK-26755] : Addressing Reviews July 17, 2019

Modifying comment in Scaladoc style and combining two unit tests into one

pgandhi999 commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -1054,6 +1043,19 @@ private[spark] object TaskSetManager {
+// (2): Tasks may be re-added to these lists multiple times as a result
+// of failures.
+// Duplicates are handled in dequeueTaskFromList, which ensures that a
+// task hasn't already started running before launching it.
```
Done

SparkQA commented:
**[Test build #107790 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107790/testReport)** for PR 23677 at commit [`7b23ef3`](https://github.com/apache/spark/commit/7b23ef37d0b38b6ee0623d894e5a9b5dbb2d4049).

SparkQA commented:
**[Test build #107790 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107790/testReport)** for PR 23677 at commit [`7b23ef3`](https://github.com/apache/spark/commit/7b23ef37d0b38b6ee0623d894e5a9b5dbb2d4049).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107790/
Test PASSed.

pgandhi999 commented on core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:
```diff
@@ -1691,79 +1692,74 @@ class TaskSetManagerSuite extends SparkFunSuite with LocalSparkContext with Logg
 
-  test("SPARK-26755 Ensure that a speculative task obeys the original locality preferences") {
-    sc = new SparkContext("local", "test")
+    // Launch a new set of tasks with locality preferences
```
I see, I apologize for misunderstanding your original intent. The thing was that I wanted to test with one set of tasks created with no locality preferences and the other set of tasks created with locality preferences, so I ended up creating two of them. I guess the best thing would then be to create two independent tests as I am not sure how do I test both cases with one taskset. Would appreciate your suggestions. Thank you.

Commit:
[SPARK-26755] : Separating unit tests and adding comment

SparkQA commented:
**[Test build #107797 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107797/testReport)** for PR 23677 at commit [`7b23b37`](https://github.com/apache/spark/commit/7b23b377b9f4c88f4bbb440f70f93b7fec776243).

pgandhi999 commented on core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:
```diff
@@ -1691,79 +1692,74 @@ class TaskSetManagerSuite extends SparkFunSuite with LocalSparkContext with Logg
-    assert(manager.resourceOffer("exec3", "host3", NODE_LOCAL).isDefined)
-    assert(manager.resourceOffer("exec4", "host4", ANY).isDefined)
+    assert(manager2.resourceOffer("exec4", "host4", NODE_LOCAL).isEmpty)
+    assert(manager2.resourceOffer("exec2", "host2", NODE_LOCAL).isEmpty)
```
Done

SparkQA commented:
**[Test build #107797 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107797/testReport)** for PR 23677 at commit [`7b23b37`](https://github.com/apache/spark/commit/7b23b377b9f4c88f4bbb440f70f93b7fec776243).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107797/
Test PASSed.

squito commented:
lgtm

I am going to be unreachable next week, so I am going to wait till I'm back to merge it (though another committer is free to merge in the meantime of course).  Sorry I know this has been delayed a lot already, but this is too brittle a part of spark for me to merge without being around to follow up on.  thanks again for all the work on it @pgandhi999 , feel free to bug me after that.

pgandhi999 commented:
Makes complete sense @squito , thank you once again for all your time.

Ngone51 commented:
@pgandhi999 I'll do another look tomorrow :)

Ngone51 commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -330,128 +294,67 @@ private[spark] class TaskSetManager(
+    // Tries to schedule a regular task first; if it returns None, then schedules
+    // a speculative task
+    dequeueTaskHelper(execId, host, maxLocality, false).orElse(
+      dequeueTaskHelper(execId, host, maxLocality, true))
```
Do `speculatableTasks -= index` when this returns `Some(index)` ?

pgandhi999 replied:
Good catch, fixed it.

Ngone51 commented on core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala:
```diff
@@ -1166,7 +1168,7 @@ class TaskSetManagerSuite extends SparkFunSuite with LocalSparkContext with Logg
     // killed, so the FakeTaskScheduler is only told about the successful completion
     // of the speculated task.
-    assert(sched.endedTasks(3) === Success)
+    assert(sched.endedTasks(4) === Success)
```
Why this changes to `4` ?

pgandhi999 replied:
We have changed the previous HashSet to ArrayBuffer, so now the dequeue logic remains same but order of tasks stored in the ArrayBuffer will be different and as the dequeuing of tasks picks up the first possible task that satisfies all conditions, a different task will be dequeued now.

Ngone51 replied:
make sense.

Commit:
[SPARK-26755] : Remove task index from speculatableTasks HashSet when a speculatable task is dequeued

SparkQA commented:
**[Test build #107922 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107922/testReport)** for PR 23677 at commit [`32fa4d2`](https://github.com/apache/spark/commit/32fa4d20587d2ca70422f3e4b50bcec68c372156).

SparkQA commented:
**[Test build #107922 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107922/testReport)** for PR 23677 at commit [`32fa4d2`](https://github.com/apache/spark/commit/32fa4d20587d2ca70422f3e4b50bcec68c372156).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107922/
Test PASSed.

xuanyuanking commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -234,63 +214,41 @@ private[spark] class TaskSetManager(
-
-  /**
-   * Return the pending tasks list for a given executor ID, or an empty list if
-   * there is no map entry for that host
```
I think we can move these three methods and comment into `PendingTasksByLocality`, then there will be fewer code changes from the original `dequeueTask` to `dequeueTaskHelper`.

pgandhi999 replied:
Not sure if we really need those three helper methods anymore, might lead to unnecessary code duplication.

xuanyuanking commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -131,37 +131,17 @@ private[spark] class TaskSetManager(
   private[scheduler] val speculatableTasks = new HashSet[Int]
 
+  // Store speculatable tasks by locality preferences
+  private[scheduler] val pendingSpeculatableTasks = new PendingTasksByLocality()
```
if the `speculationEnabled` is false, we can even don't do the instance creation here.

pgandhi999 replied:
We were not doing this earlier, so not sure whether we should do it now.

xuanyuanking commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -302,16 +260,23 @@ private[spark] class TaskSetManager(
+        if (!successful(index)) {
+          if (copiesRunning(index) == 0) {
+            return Some(index)
+          } else if (speculative && copiesRunning(index) == 1) {
```
So here is the logic of speculating task only once? How about adding more comments to highlight.

pgandhi999 replied:
Added comment, thank you.

xuanyuanking reviewed (commented):
+1 for both better performance and code reusing.

Ngone51 commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -302,16 +260,23 @@ private[spark] class TaskSetManager(
+          if (copiesRunning(index) == 0) {
+            return Some(index)
+          } else if (speculative && copiesRunning(index) == 1) {
+            speculatableTasks -= index
```
when `copiesRunning(index) == 0`, it is still possible to return a speculatable tasks, thinking about the case that a long running task failed after it got a pending  speculatable task. I mean,  for `if (copiesRunning(index) == 0)` above, it actually covers two cases:  1) not speculative 2) speculative. So, I think it's better to put this after `dequeueTaskHelper(execId, host, maxLocality, true)`.

pgandhi999 replied:
Makes sense, have updated the code.

Commit:
[SPARK-26755] : Addressing Reviews July 22, 2019

Removing task indes from speculatableTasks after task is dequeued, adding comment

SparkQA commented:
**[Test build #108019 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108019/testReport)** for PR 23677 at commit [`685bbae`](https://github.com/apache/spark/commit/685bbae298eb84de4282b690fd410d8404e6c424).

SparkQA commented:
**[Test build #108019 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108019/testReport)** for PR 23677 at commit [`685bbae`](https://github.com/apache/spark/commit/685bbae298eb84de4282b690fd410d8404e6c424).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108019/
Test PASSed.

Ngone51 commented:
LGTM

squito commented on core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala:
```diff
@@ -107,7 +107,7 @@ class OutputCommitCoordinatorSuite extends SparkFunSuite with BeforeAndAfter {
       new TaskSetManager(mockTaskScheduler, taskSet, 4) {
         private var hasDequeuedSpeculatedTask = false
-        override def dequeueSpeculativeTask(execId: String,
+        def dequeueSpeculativeTask(execId: String,
```
sorry I hadn't noticed this before -- this change is wrong as now this whole setup isn't doing anything.  To keep the same behavior as before, I think you need to change `dequeueTaskHelper` to be `protected`, and then override it here.

https://github.com/squito/spark/commit/366e916f4e7499d477d6fafda538d5fe7d65d6a4

pgandhi999 replied:
Makes sense, have updated the code. Sorry, I missed out that one earlier. Thanks @squito.

Commit:
[SPARK-26755]: Merge branch 'master' of https://github.com/pgandhi999/spark into SPARK-26755

Upmerging with master branch

Commit:
[SPARK-26755] : Addressing Reviews July 29, 2019

Fixing OutputCommitCoordinatorSuite Test to override dequeueTaskHelper method.

SparkQA commented:
**[Test build #108346 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108346/testReport)** for PR 23677 at commit [`7a8c992`](https://github.com/apache/spark/commit/7a8c9926137ea527ec9d0bf1ad170dadf5e92c0d).

SparkQA commented:
**[Test build #108346 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108346/testReport)** for PR 23677 at commit [`7a8c992`](https://github.com/apache/spark/commit/7a8c9926137ea527ec9d0bf1ad170dadf5e92c0d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108346/
Test PASSed.

squito reviewed (approved):
lgtm

squito commented:
merged to master, thanks @pgandhi999 !

pgandhi999 commented:
Thank you @squito and @Ngone51 for all your valuable reviews.
